tensorflow_addons/losses/python/lifted.py

```diff
@@ -0,0 +1,141 @@
+
+  # Build pairwise binary adjacency matrix.
+  adjacency = math_ops.equal(labels, array_ops.transpose(labels))
+  # Invert so we can select negatives only.
+  adjacency_not = math_ops.logical_not(adjacency)
+
+  batch_size = array_ops.size(labels)
+
+  diff = margin - pairwise_distances
+  mask = math_ops.cast(adjacency_not, dtype=dtypes.float32)
+  # Safe maximum: Temporarily shift negative distances
+  #   above zero before taking max.
+  #     this is to take the max only among negatives.
+  row_minimums = math_ops.reduce_min(diff, 1, keepdims=True)
+  row_negative_maximums = math_ops.reduce_max(
+      math_ops.multiply(diff - row_minimums, mask), 1,
+      keepdims=True) + row_minimums
+
+  # Compute the loss.
+  # Keep track of matrix of maximums where M_ij = max(m_i, m_j)
+  #   where m_i is the max of alpha - negative D_i's.
+  # This matches the Caffe loss layer implementation at:
+  #   https://github.com/rksltnl/Caffe-Deep-Metric-Learning-CVPR16/blob/0efd7544a9846f58df923c8b992198ba5c355454/src/caffe/layers/lifted_struct_similarity_softmax_layer.cpp  # pylint: disable=line-too-long
+
+  max_elements = math_ops.maximum(
+      row_negative_maximums, array_ops.transpose(row_negative_maximums))
+  diff_tiled = array_ops.tile(diff, [batch_size, 1])
+  mask_tiled = array_ops.tile(mask, [batch_size, 1])
+  max_elements_vect = array_ops.reshape(
+      array_ops.transpose(max_elements), [-1, 1])
+
+  loss_exp_left = array_ops.reshape(
+      math_ops.reduce_sum(
+          math_ops.multiply(
+              math_ops.exp(diff_tiled - max_elements_vect), mask_tiled),
+          1,
+          keepdims=True), [batch_size, batch_size])
+
+  loss_mat = max_elements + math_ops.log(
+      loss_exp_left + array_ops.transpose(loss_exp_left))
+  # Add the positive distance.
+  loss_mat += pairwise_distances
+
+  mask_positives = math_ops.cast(
+      adjacency, dtype=dtypes.float32) - array_ops.diag(
+          array_ops.ones([batch_size]))
+
+  # *0.5 for upper triangular, and another *0.5 for 1/2 factor for loss^2.
+  num_positives = math_ops.reduce_sum(mask_positives) / 2.0
+
+  lifted_loss = math_ops.truediv(
+      0.25 * math_ops.reduce_sum(
+          math_ops.square(
+              math_ops.maximum(
+                  math_ops.multiply(loss_mat, mask_positives), 0.0))),
+      num_positives,
+      name='liftedstruct_loss')
+  return lifted_loss
+
+
+@keras_utils.register_keras_custom_object
+class LiftedStructLoss(losses.LossFunctionWrapper):
+    """Computes the lifted structured loss.
+
+    The loss encourages the positive distances (between a pair of embeddings
+    with the same labels) to be smaller than any negative distances (between a
+    pair of embeddings with different labels) in the mini-batch in a way
+    that is differentiable with respect to the embedding vectors.
+    See: https://arxiv.org/abs/1511.06452.
+
+    Args:
+        labels: 1-D tf.int32 `Tensor` with shape [batch_size] of
+        multiclass integer labels.
```

Comment:
Please make sure all doc format are consistent, refer to https://github.com/tensorflow/addons/pull/28/files

Reply:
Do you want it in a single line